### ISSUE TITLE
Closing the application when any tracker is loaded results in a crash

### ISSUE THREAD

**Issue by justayak** [BioroboticsLab/biotracker_core#60]
Closing the application when any tracker is loaded results in a crash that prevents the destructors of various objectes (including the BioTrackerApp main object and the tracker object itself) to be called.

```bash
*** Error in `./BioTracker': double free or corruption (out): 0x000000000293e660 ***
[2]    3119 abort (core dumped)  ./BioTracker
```

When the trackers are switched out in the GUI the dtors are called as expected.

**Comment by justayak** [comment]
The problem occurs when a tracker has a an std::shared_ptr<QWidget> as its member variable and returns it to the user interface in the "getToolsWidget" call. A quick fix is to not have a member variable but instead return a std::shared_ptr<QWidget> that was created inside tof "getToolsWidget". However, this needs further investigation.

**Comment by justayak** [comment]
I would even argue that we do not need a shared_ptr but rather return the QWidget object itself as it will be (and must be) added to a qt gui element which, in turn, will ensure that the object will be deleted when needed. To ensure memory leaks, we then would need to manually remove the QWidget from the parent when the tracker is being destroyed. All this logic can be handled in the abstract Tracker so that it would be *invisible* for people who implement their own trackers

**Comment by justayak** [comment]
Current suggestion: The abstract TrackingAlgorithm will hold a pointer to a QWidget *m_tools* but will NOT hold the ownership. Furthermore, getToolsWidget will NOT be virtual anymore but rather just return m_tools. Instead, trackers need to instantiate their tools in their ctors with the abstract TrackingAlgorithm`s *m_tools* as their parent. In the Registry, upon switching out a tracker, we *delete* the m_tools object of the old tracker.

**Comment by TobiasLudwig** [comment]
This sounds way to complicated. Could we just remove the widget from the layout (<code>m_tools.removeWidget(m_lastToolsWidget.get());</code>) (and reset it's parent?) in the <code>MainWindow</code>'s destructor to prevent the duplicate destruction and don't transfer ownership? 
I agree the the std::shared_ptr is superfluous. A raw pointer that is internally managed by a std::unique_ptr is sufficient.

**Comment by justayak** [comment]
The issue here is that we want the code at the *client* side to be as easy to use as possible as the core part now poses as a library that should hide most of the complicated stuff. With the implementation I mentioned  above we would achieve that: the gui logic would actually become much simpler as it does not need to take care of removing the old widgets anymore as well while the widget will be just free`d once.
On a side node regarding the removal of the *getToolsWidget()* from the extending tracker: I had a rather involved discussion with @nebw about it and in the end he convinced me that its best in cpp to initialize all member variables, which would include potential gui elements, in the constructor rather than in a  deferred function - which now seems very obvious to me as well..

-----
in short: Code for the GUI and for the trackers actually becomes simpler while the code in the core becomes slightly more complex - which is a good tradeoff given that we want the api of the Core lib to be nice..

**Comment by TobiasLudwig** [comment]
If I get you right the suggested approach leaks the tools Object if the tracker is used on its own:
```
struct Foo : TrackingAlgorithm {
  Foo() {
    m_tools = new ...
  }
};

void bar() {
  Foo f;
}
```
I believe the double free issue could also be solved by changing the destruction order: I suppose that currently the GUI is destroyed first and thus the tools widget. Therefore the tracker's shared_ptr now holds an invalid pointer. Destroying the tracker first should remove the widget from the GUI's layout.

**Comment by justayak** [comment]
You are right. However, the Tracker is only supposed to reside in the context of the BioTracker Core, which, in turn, will take care of the widget object. This is ugly, but it will be the responsibility of our framework to take care of this issue rather than any potential user of our library (like the gui). It would be really inconvenient for any user of the core lib to ensure a specific destruction order.. thus I would still opt for our solution

**Comment by TobiasLudwig** [comment]
If it's possible to create a Tracker on it's own someone will eventually do so. This approach also complicates unit testing a Tracker. A `QPointer`might solve this:
```
class TrackingAlgorithm {
private:
QPointer<QWidget> m_tools;
public:
~TrackingAlgorithm() {
  delete m_tools;
}
void setTools(QWidget *w) {
  delete m_tools;
  m_tools = w;
}
QWidget* getTools() {
  return m_tools;
}
};
```

**Comment by justayak** [comment]
*QPointer* is the solution: it can detect if a deletion happened and act accordingly. That will be the best solution. Thanks @TobiasLudwig
